File: go_playground/src/algorithm/knapsack.go
package algorithm

func Max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func Knapsack(wt, val []int, n, w int) int {
	dp := [100][100]int{}
	for i := 1; i <= n; i++ {
		for j := 1; j <= w; j++ {
			if j >= wt[i-1] {
				dp[i][j] = max(dp[i-1][j], dp[i-1][j-wt[i-1]]+val[i-1])
			} else {
				dp[i][j] = dp[i-1][j]
			}
		}
	}
	return dp[n][w]
}

File: rust_playground/src/dp/knapsack.rs
pub struct Knapsack;

impl Knapsack {
    pub fn knapsack(wt: Vec<usize>, val: Vec<usize>, n: usize, w: usize) -> usize {
        let mut dp = vec![vec![0; w + 1]; n + 1];
        for i in 1..=n {
            for j in 1..=w {
                dp[i][j] = if j < wt[i - 1] {
                    dp[i - 1][j]
                } else {
                    dp[i - 1][j].max(dp[i - 1][j - wt[i - 1]] + val[i - 1])
                }
            }
        }
        dp[n][w]
    }
}

//// tests
///

#[cfg(test)]
mod test {
    use super::Knapsack;

    #[test]
    fn test_knapsack() {
        assert_eq!(Knapsack::knapsack(vec![1, 2, 3], vec![3, 2, 1], 2, 4), 5)
    }
}
